### PR TITLE
ENH: Add ellipse class for annotation box styles

### LINKED ISSUE #24568: [ENH]: Ellipse annotation
### Problem

Currently there is only a circle annotation, which can be quite big for long texts.

### Proposed solution

Add an ellipse annotation. Equations can be found here (second answer): https://stackoverflow.com/questions/433371/ellipse-bounding-a-rectangle

The code will be very similar to the circle annotation:
https://github.com/matplotlib/matplotlib/blob/dc0328fb9c1c8817b07a5121a106a06be3571433/lib/matplotlib/patches.py#L2327-L2346

but with an ellipse rather than a circle.

There is also a bit of documentation to update, e.g., https://matplotlib.org/stable/tutorials/text/annotations.html

### PR DESCRIPTION
**Pull request by ckcherry23**
## PR Summary
Create a new 'Ellipse' annotation box style that is modelled by the 'Circle' class
Fixes #24568 

## PR Checklist

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] New plotting related features are documented with examples.

**Release Notes**
- [x] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [x] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/` [N/A]
- [x] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by oscargus** [comment]
Looks promising!

If you can add info here:
https://github.com/matplotlib/matplotlib/blob/7f7dfc9303326c283593270cfcdb1b439781563a/tutorials/text/annotations.py#L202-L203
of the new shape.

Here one can see that it is working:
image

I'll add some comments in the code as well.

There is a failing test at the moment. It should just be to update the test image so that includes the ellipse as well.

File: lib/matplotlib/patches.py
Comment (oscargus) [inline lib/matplotlib/patches.py:2345]:
```suggestion
        """
        An elliptical box.
        
        .. versionadded:: 3.7
        """
```

File: lib/matplotlib/patches.py
Comment (oscargus) [inline lib/matplotlib/patches.py:2366]:
```suggestion
            return trans.transform_path(Path.unit_circle())
```
for brevity.

**Comment by oscargus** [comment]
This is the current test image:
image

which will now also include the ellipse. So updating that image after rerunning the test, copying from result_images to baseline_images will have the tests pass and the new code tested.

**Comment by oscargus** [comment]
Oh, and I think that it may be worthwhile to highlight this addition through a What's new note. If you want to, write some short text and a minor example showing how it can be used. See examples is https://github.com/matplotlib/matplotlib/tree/main/doc/users/next_whats_new

File: doc/users/next_whats_new/ellipse_annotation.rst
Comment (tacaswell) [inline doc/users/next_whats_new/ellipse_annotation.rst:4]: These are causing:

```
/home/circleci/project/doc/users/next_whats_new/ellipse_annotation.rst:4: WARNING: py:obj reference target not found: ellipse
/home/circleci/project/doc/users/next_whats_new/ellipse_annotation.rst:4: WARNING: py:obj reference target not found: circle
```

Sphinx using single-back tick to identify things that need to be cross-referenced.  I think the correct fix is to change these to ``'ellipse'`` and ``'circle'``.

**Review by tacaswell — APPROVED** [review]
modulo fixing the docs build.

**Comment by oscargus** [comment]
Agree that it looks good! Let me just check the equations (now it seems like the padding is scaled, but I cannot really get my head around it as it looks like the padding is scaled down by sqrt(2), but in the image it looks like the padding has increased if anything).

**Comment by oscargus** [comment]
OK, there is a slightly different behavior of the `pad` here, but I am not sure if it is a problem or not. Nor if it is worthwhile to redo the code.

Here is what happens with circle, this and what possibly may be a preferred interpretation of `pad`:

image

As seen, for circle, the leftmost, the padding is on the radius of the circle. For this PR, the middle, the padding is on the bounding box and then the ellipse is touching the padded box. One may think that the padding could be as in the rightmost figure, but I do not think that it is a problem as one probably will try to figure out a good padding anyway.

(I also realized that with `pad=0` a circle does not always enclose all the text. One can try this with `plt.text(0.5, 0.5, "M", bbox={'pad': 0})`.)

I'll approve this as the code is good as such, but not merge it to get a second opinion on the behavior.

Thanks for your effort!

**Review by oscargus — APPROVED** [review]
Good code, but would like to have a second opinion on the behavior.

**Comment by tacaswell** [comment]
- Pro of just merging now: get it in!
- con of just merging now: may have to regenerate a test image

**Comment by jklymak** [comment]
@oscargus how are you defining the angle on the pad in the above?  Is it going through the center of the text bounding box?

**Comment by oscargus** [comment]
That was my idea, yes, as that seems to be a quite natural way to define it, closest distance to the bounding box (although any angle will have that...). But I guess that there are more options, like 45 degrees.

Considering that the Circle pad is a bit different compared to the others, I am not sure if it actually matters that Ellipse is a bit "off" as well. I guess my main idea was that the pad should be visually similar for the same value.

**Comment by oscargus** [comment]
Another reason is that the equations become a bit easier for that case, compared to e.g. 45 degrees, although the computational burden is not that huge anyway.

**Comment by tacaswell** [comment]
On a bit more consideration I think that there are a bunch of reasonable definitions of pad and there is really only one degree of freedom here.   It is not clear to me one definition of pad is more obviously right or more natural nor is it clear that the definition of `pad` should be consistent between different annotations styles (I expect these to be tuned by what "looks good' and I really hope that people do not try to encode quantitative information into the size of the annotation boxes over what is needed to fit the text....).  In addition to the two proposal I could also see pad being what is needed for either the major or minor axis to be tangent to the sides of text size (which seems to be the generalization of `Circle`'s pad?).

I'm in favor of merging this as-is and accept the apparent inconsistency because the inconsistency is inevitable.

**Comment by oscargus** [comment]
Thanks @ckcherry23 ! Hope to see you around!

**Pull request merged**
